crypto/sm2/sm2_sign.c

```diff
@@ -25,16 +25,17 @@ static BIGNUM *sm2_compute_msg_hash(const EVP_MD *digest,
 {
     EVP_MD_CTX *hash = EVP_MD_CTX_new();
     const int md_size = EVP_MD_size(digest);
-    uint8_t *za = OPENSSL_zalloc(md_size);
+    uint8_t *za;
     BIGNUM *e = NULL;
 
-    if (hash == NULL || za == NULL) {
-        SM2err(SM2_F_SM2_COMPUTE_MSG_HASH, ERR_R_MALLOC_FAILURE);
+    if (md_size < 0) {
```

Comment:
This is effectively impossible condition. In sense that if `digest` pointer is real, then it would be indication of severe memory corruption. Well, so would be `digest` pointer being unreal. Bottom line is that ossl_assert would be appropriate, so that debug build would crash...

Reply:
> it could come back with a -1 result if digest is NULL

Or it could come back as -13 if digest pointer was corrupted. Note that suggestion is not to remove the check, but replace it with ossl_assert, as debugging aid.

Reply:
My point is that I'm not convinced ossl_assert is appropriate. ossl_assert() is for internal errors. Depending on how this is exposed in an API in response to #6534 it could occur as a result of an application programming error.

Reply:
Well, I don't see the conflict. I mean how does assumption about application programming error disqualifies ossl_assert... But whatever...

Reply:
As far as programming errors go, I'd say that it would be more appropriate to check for digest to be NULL and push error that says "no digest passed". And it might be more informative if it's done closer to caller. I mean if caller calls function X, message would be more readable if it spells "you have to pass digest to function X" and not "digest was NULL in Y, which in invalid for Z".

Reply:
> As far as programming errors go, I'd say that it would be more appropriate to check for digest to be NULL and push error that says "no digest passed". And it might be more informative if it's done closer to caller.

Possibly, but even if we did that then I'd still say we need to check for a negative response from EVP_MD_size(). The external behaviour of that function is that it returns a negative result on error - not that it returns a negative result on a NULL input. The fact that that is the only error we currently have is not the point. We reserve the right to add new errors in the future. We would expect applications to be able to handle it if we introduced a new error. Our own code should be the same.

As for doing it closer to the caller, then I'd agree. However, until #6534 is addressed we can't do that.

Reply:
Well, as far as specifically EVP_MD_size goes, one can probably argue both ways. I mean it's just a getter, and there is a bunch of similar getters. *None* of them check for argument being NULL. So why EVP_MD_size would be different? Just because it's there doesn't necessarily make it right. It might have been added in ad-hoc manner, for example in attempt to appease coverity... Even though this is side note in the context, one can [and probably should] wonder if generalization such as "negative result on error - not on specifically NULL input" is actually justified...

Reply:
> it's just a getter

And not just getter, but getter for constant structure. Constant not in sense that EVP_MD_size argument is declared as constant pointer, but in sense that object is supposed to remain immutable for its life time.